README
This is cfunge - a fast Befunge93/98/109 interpreter in C.

cfunge offers some features that many other standard conforming interpreters
don't. For example:
 * Sandbox mode, prevents programs from harming system (more details below).
 * Advance tracing support, debugging protocol support system under development.
 * Passes mycology (of course other conforming interpreters does this, but most
   interpreters are, sadly, not conforming).
 * Tested using Valgrind and similar tools.
 * Tested with fuzz testing to ensure cfunge does not segfault on random stuff.
 * Coded for maximum performance AND correctness.
 * Support for selecting either 64-bit or 32-bit integers as datatype in
   funge-space at compile time.


Dependencies
------------
REQUIRED:
 * cmake (http://www.cmake.org/) to generate Makefile for cfunge. At least
   version 2.6.3 is required.

 * A C compiler, or one that supports a large subset of C99, such as GCC.
   + GCC 3.4.6, 4.1.2, 4.2.1, 4.3.2 and 4.3.3 are known to work, other versions
     may or may not work.
   ! Warning: GCC 4.5 and 4.6 breaks with -O3 currently.
     (Later versions unknown.)
   + ICC 10.1 and 11.1 are known to work too.
   + Open64 4.1 should work, but isn't recommended.
   + clang 1.0 (LLVM 2.6) and clang 2.9 (LLVM 2.9) are known to work.
   + PathScale EKOPath 4.0.10 is known to work.
   - TCC 0.9.24 is known to fail at certain C99 constructs used in cfunge.
   Other compilers may or may not work.

 * A POSIX.1-2001 system with the memory mapped file option. It also needs the
   function strdup() which is part of the XSI extension. (For POSIX.1-2008 both
   strdup() and memory mapped files are mandatory and part of the base.)
   Operating systems known to work:
   + Linux 2.6.28 (with glibc 2.9) and later (earlier versions unknown but will
     probably work unless very old).
   + FreeBSD 6.4 or later (earlier versions unknown).
   + Mac OS X (unknown version) has been reported to work but I don't have it
     myself and thus can't test it.
   + NetBSD 5.0.1/amd64 is known to work.
   + OpenBSD 4.4 (sparc64): I have not managed to get cmake 2.6 to compile on
     OpenBSD, but I did manage to build cfunge by hand, required some strange
     workarounds and defines.
   Windows most likely won't work. Nor do I plan to support it. Cygwin may work,
   but you are on your own if it doesn't.

HIGHLY RECOMMENDED:
 * Ncurses (http://www.gnu.org/software/ncurses/), needed for the TERM
   fingerprint. This is most likely already installed. Ncurses will be
   automatically used if found.

 * IEC 60559 floating-point arithmetic. Please see Annex F in ISO/IEC 9899 for
   more details.


Configuring
-----------
Warning: Out of tree builds are highly recommended. Building in the source tree
may work but is _completely_ untested.

Warning: Only the make and ninja backends for cmake have been tested. Other
backends may work, but are untested.

To build using cmake, you can run these commands in the top source directory:
  mkdir build && cd build
  cmake ..

If you are on a 32-bit system you may want to use 32-bit integers (instead of
64-bit integers) for speed. It is also slightly faster on some 64-bit systems
to use 32-bit integers. This may vary between architectures. To use 32-bit
integers you could use these commands instead of the above ones:
  mkdir build && cd build
  cmake -DUSE_64BIT=OFF ..

If you want to see a list of available options use ccmake. It will allow you to
select options in a ncurses based user interface. Help is always shown at the
bottom of the screen.
  mkdir build && cd build
  ccmake ..
    (press c)
    (change options - use t to show advanced options)
    (press c again)
    (press g to generate make file)

For more information see:
  cmake --help
and/or
  ccmake --help


Compiling
---------
After having run cmake as described in the above section, just run:
  make


Installing
----------
Not needed, cfunge can be run from build directory, but if you want to (after
having compiled cfunge):
  make install


Fingerprints
------------
It is planned to implement most or all of the existing fingerprints,
with some exceptions:
 * FNGR - Contradicts with 98 standard.
 * IMAP - Too intrusive.
 * MODE - Intrusive into IP handling.
 * TRDS - Exceedingly complex and intrusive.
 * WIND - Too complex to implement and not portable.

Short descriptions of implemented fingerprints:
 3DSP 3D space manipulation extension
 BASE I/O for numbers in other bases
 BOOL Logic Functions
 CPLI Complex Integer extension
 DATE Date Functions
 DIRF Directory functions extension
 FILE File I/O functions
 FING Operate on single fingerprint semantics
 FIXP Some useful math functions
 FPDP Double precision floating point
 FPSP Single precision floating point
 FRTH Some common forth commands
 HRTI High-Resolution Timer Interface
 INDV Pointer functions
 JSTR Read and write strings in Funge-Space
 MODU Modulo Arithmetic Extension
 NCRS Ncurses Extension
 NULL Funge-98 Null Fingerprint
 ORTH Orthogonal Easement Library
 PERL Generic Interface to the Perl Language
 REFC Referenced Cells Extension
 REXP Regular Expression Matching
 ROMA Funge-98 Roman Numerals
 SCKE TCP/IP async socket and dns resolving extension
 SOCK TCP/IP socket extension
 STRN String functions
 SUBR Subroutine extension
 TERM Terminal control functions
 TIME Time and Date functions
 TOYS Funge-98 Standard Toys
 TURT Simple Turtle Graphics Library
For more details please see the specs for each fingerprint.
In cases of undefined behaviour in fingerprints, cfunge mostly tries to do the
same thing as CCBI.


Undefined behaviour
-------------------
The Befunge98 standard leaves some things undefined, here is what we do for
some of those cases:
 * y pushes time in UTC not local time.
 * k with a negative argument reflects.
 * # across edge of funge-space may or may not skip first char after wrapping
   depending on exact situation.
 * ( and ) with a negative count reflects and doesn't pop any fingerprint.
 * Loaded fingerprints are inherited to child IPs at split (t).
 * STDOUT is only flushed at:
   * Newline (line feed, ASCII 10) printed using , instruction.
   * Any input instructions.
   * End of program.
 * Standard input is read one line at a time and buffered internally. Those
   instructions reading chars fetch one char from this buffer, leaving the rest
   (if any) including any ending newline. Instructions reading an integer will
   leave anything after the integer in the buffer with one exception: if the
   next char is a newline it will be discarded.


Notes on different standards
----------------------------
The option -s 93 does not prevent the program from accessing outside the first
80x25 cells. Nor does it disallow instructions that didn't exist in 93. It does
however change space behaviour to match 93 style.

If a program depends on a instruction that is undefined in 93 to reflect, it
should be easy to replace such instructions with a r for reflect or any in the
range A-Z (and not load any fingerprint).

Further division by zero always returns 0 in all modes, though the Befunge93
specs says the interpreter should ask the user what result he/she wants in that
situation.


Sandbox mode
------------
Sandbox mode prevents Funge programs from doing "harmful" things, this includes,
but is not limited to:
 * Any file or filesystem IO is forbidden.
 * The list of environment variables the program can see in y are restricted.
 * Non-safe fingerprints can not be loaded (this includes network and file
   system access as well as other things).


ChangeLog
ChangeLog
=========
This is a overview of changes users may care about. Detailed changelog can be
done using bzr (does not work in tarballs, you need a development checkout for this):
  bzr log -rtag:tag.. --short
Example (list changes compared to 0.1.0)
  bzr log -rtag:0.1.0.. --short
For more info:
  bzr help log
and
  bzr help revisionspec


Development version (up to r883)
--------------------------------
New features and improvements:
 * Added alternative IP list implementation, that stores pointers to the IPs
   instead. Useful for very large number (thousands) of IPs. This is now the
   default since the performance drop with a small number of threads is tiny
   compared to the increase for a large number of threads.
 * Cache env vars and other constant (per execution) data in y.
 * Changed stack stack code to not realloc once for each call of { and }.
 * Improved speed for non-cardinal warp.
 * Made cfunge work with the PathScale EKOPath compiler.

Changed features:
 * Dropped support for Boehm-GC. It was not well tested and manual memory
   management worked better.
 * Removed scaling of coordinates from TURT.
 * Dropped inline asm since intrinsics work well on modern GCC and ICC.
 * Improved fuzz testing script.

Major bug fixes:
 * Fixed bugs in run_sys_info in stack_get_index that together caused
   incorrect behaviour for y picking of top of stack and for last 0 below
   env vars.
 * Removed counts from hash tables if they reach zero. Needed for large model
   funge-space bounds shrinking logic to be correct.
 * Fixed bug in STRN G that resulted in an extra \0 being pushed.
 * Fixed STRN G, it incorrectly subtracted 1 from string length when
   calling stack_push_string().
 * Fixed several STRN functions truncating strings to unsigned chars.

Minor bug fixes:
 * Fixed possible problem in FRTH for very large stack sizes (outside range
   of funge cells, but inside range of size_t).
 * Various small fixes for clang and ICC.
 * Various build system fixes.
 * Handle zero bytes properly in stdin. Note that STRN still retains the old
   behaviour and arguably should do so.
 * Fixed various bugs (including memory leaks) when out of memory.
 * Fixed bug in text mode file output for lines with a single (non-space)
   char on them.
 * Add missing NULL pointer checks in code for i, o and = as well as in several
   fingerprints.

Added fingerprints:
 * BOOL Logic Functions


0.9.0
-----
This is a bug fix release. The version number changed to 0.9.0 to reflect that
cfunge is now reaching a stable version.
 * Bug fix: In a file with CR line endings and lines with leading spaces, the
   leading spaces were lost.
 * Bug fix: When j was used to jump backwads (negative argument) to a cell
   directly on the edge of the current bounding box of the Funge Space, it would
   incorrectly wrap. Fix this by peforming the normal moving forward of the IP
   before the jump in j instead of after as usual for most other instructions.
 * Minor bug fix: Hypens were incorrectly escaped when generating man pages in
   tools/gen_fprint_list.sh. Fix and update man page to reflect.


0.4.1
-----
Major highlights:
 * cfunge can now track exact bounds for Funge Space (option at compile time).
 * Support for disabling some heavy parts of cfunge to reduce memory usage (see
   below for details).
Other important changes since last release:
 * For exact bounds: Implement sparse bounds minimising when difference is huge.
 * Reworked sysinfo code, now less overhead and more reused results (from time()
   and fungespace_get_bounds_rect()).
 * Bug fix: y pushed bounds wrong in certain cases.
 * Bug fix: Fixed OOM behaviour in cfunge on huge count for stack-stack begin.
 * DATE: Make sure constants are long double ones (L suffix). Do more of the
         conversion as integer math to reduce rounding errors. Make sure to
         always use int_least64_t internally for greatest precision. Add some
         missing function attributes.
 * TURT and FIXP: Use more exact value of pi than M_PI for long double
   calculations.
 * Added system for diagnostics output, collected in one place. This should make
   the error messages more uniform.
 * Added new command line switch (-E) to show non-fatal errors.
 * Added the command line switch (-v) to display build info and features in a
   compact format.
 * Reordered some code in file loading to make sure we don't push uninitialised
   size in i when trying to load an empty file.
 * Removed unused functions (cf_strndup() and cf_strnlen()).
 * Cleaned up includes (removing unused includes in header and source files).
 * Avoid malloc() calls in BASE. The max needed size is small and thus we can
   safely use a VLA.
 * Avoid some malloc() calls for fingerprint opcode stacks when they are empty.
   Now the actual stacks are not allocated until they are needed. Also "inline"
   the opcode stack metadata into the ip structs to reduce memory overhead.
 * Added the file doc/API_CHANGES documenting API changes. Please see that file
   for changes in the API.
 * Fixed build issue with IFFI.
 * Fixed handling of SSE intrinsics and inline asm for ICC in funge-space.c.
 * Build system:
   * Cleaned up warning flags in CMakeLists.txt.
   * Add options to disable some optional features even if needed dependencies
     found. Features that can be disabled are: clock_gettime(), ncurses,
     fingerprints using floating point, TURT.
   * Clean up CMakeLists.txt splitting some stuff out into separate files.


0.4.0
-----
This release add support for some more fingerprints, speed up execution quite
a bit, and fixes several bugs. See details below.

Important changes in this release:
 * cfunge now depend on cmake 2.6 (or later).
 * cfunge is now quite a bit faster thanks to many parts rewritten in a faster
   way. This includes:
     * Stack-stack begin/end are faster.
     * Pushing strings on stack have been optimised a lot (important for y and
       various fingerprints).

src/main.c
/* -*- mode: C; coding: utf-8; tab-width: 4; indent-tabs-mode: t; c-basic-offset: 4 -*-
 *
 * cfunge - A standard-conforming Befunge93/98/109 interpreter in C.
 * Copyright (C) 2008-2013 Arvid Norlander <anmaster AT tele2 DOT se>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at the proxy's option) any later version. Arvid Norlander is a
 * proxy who can decide which future versions of the GNU General Public
 * License can be used.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * @file
 * Contains command line parsing and such.
 */
#include "global.h"
#include "main.h"

#include <stdio.h>  /* fprintf, puts */
#include <stdlib.h> /* exit */
#include <signal.h> /* signal */
#include <string.h> /* strncmp */
#include <unistd.h> /* getopt */
#include <limits.h> /* CHAR_BIT */

#ifdef FUZZ_TESTING
#include <sys/resource.h>
#endif

#include "diagnostic.h"
#include "interpreter.h"
#include "settings.h"
#include "fingerprints/manager.h"

const char *const *fungeargv = NULL;
int fungeargc = 0;

// Exclude some code if we are building in IFFI.
#ifndef CFUN_IS_IFFI
// Use a larger buffer for stdout in fully buffered mode.
static char cfun_iobuf[BUFSIZ * 4];

// These are NOT worth inlineing, even though only called once.
FUNGE_ATTR_NOINLINE FUNGE_ATTR_COLD FUNGE_ATTR_NORET
static void print_features(void)
{
	puts("Features compiled into this binary:\n"
#ifdef CONCURRENT_FUNGE
	     " + Concurrency using t instruction is enabled.\n"
#else
	     " - Concurrency using t instruction is disabled.\n"
#endif

#ifndef DISABLE_TRACE
	     " + Tracing using -t <level> option is enabled.\n"
#else
	     " - Tracing using -t <level> option is disabled.\n"
#endif

#ifdef CFUN_EXACT_BOUNDS
	     " + This binary uses exact bounds in y.\n"
#else
	     " - This binary does not use exact bounds in y.\n"
#endif

#ifdef DEBUG
	     " * This binary is a debug build.\n"
#endif

#ifndef NDEBUG
	     " * This binary is compiled with asserts.\n"
#endif

#ifdef ENABLE_VALGRIND
	     " * This binary is compiled with valgrind debugging annotations.\n"
#endif

#if defined(USE64)
	     " * Cell size is 64 bits (8 bytes).\n"
#elif defined(USE32)
	     " * Cell size is 32 bits (4 bytes).\n"
#else
#  error "Unknown cell size."
#endif

	     // Features with ! are stuff most users doesn't want.
#ifdef CFUN_NO_FLOATS
	     " ! This binary is compiled without any floating point fingerprints.\n"
#endif

#ifdef FUZZ_TESTING
	     // We use this to warn users and to do sanity checking in the fuzz testing
	     // script.
	     " ! This is a fuzz testing build and thus not standard-conforming.\n"
#endif

	    ); /* End of puts() call */

	// This call does not return.
	manager_list();
}

FUNGE_ATTR_NOINLINE FUNGE_ATTR_COLD FUNGE_ATTR_NORET
static void print_help(void)
{
	puts("Usage: cfunge [OPTIONS] [FILE] [PROGRAM OPTIONS]\n"
	     "A fast Befunge interpreter in C\n\n"
	     " -b           Use fully buffered output (default is system default for stdout).\n"
	     " -E           Show non-fatal error messages, fatal ones are always shown.\n"
	     " -F           Disable all fingerprints.\n"
	     " -f           Show list of features and fingerprints supported in this binary.\n"
	     " -h           Show this help and exit.\n"
	     " -S           Enable sandbox mode (see README for details).\n"
	     " -s standard  Use the given standard (one of 93, 98 [default] and 109).\n"
	     " -t level     Use given trace level. Default 0.\n"
	     " -V           Show version and copyright info and exit.\n"
	     " -v           Show version and build info and exit.\n"
	     " -W           Show warnings."
#ifdef DISABLE_TRACE
	     "\nNote that someone disabled trace in this binary, so -t will have no effect."
#endif
	    );
	exit(EXIT_SUCCESS);
}

FUNGE_ATTR_NOINLINE FUNGE_ATTR_COLD FUNGE_ATTR_NORET
static void print_build_info(void)
{
	printf("cfunge " CFUNGE_APPVERSION " ["
#ifdef CONCURRENT_FUNGE
	       "+con "
#else
	       "-con "
#endif
#ifndef DISABLE_TRACE
	       "+trace "
#else
	       "-trace "
#endif
#ifdef CFUN_EXACT_BOUNDS
	       "+exact-bounds "
#else
	       "-exact-bounds "
#endif
#ifdef HAVE_NCURSES
	       "+ncurses "
#else
	       "-ncurses "
#endif
#ifdef _FORTIFY_SOURCE
	       "hardened "
#endif
#ifdef DEBUG
	       "debug "
#endif
#ifndef NDEBUG
	       "asserts "
#endif
#ifdef ENABLE_VALGRIND
	       "valgrind "
#endif
#ifdef _MUDFLAP
	       "mud "
#endif
#ifdef CFUN_NO_FLOATS
	       "nofloat "
#endif
#ifdef FUZZ_TESTING
	       "fuzz "
#endif
	       // Pointer size and Cell size
	       "p:%zu c:%zu]\n\n"
	       "Platform:      " CFUN_TARGET_PLATFORM "\n"
	       "OS:            " CFUN_TARGET_OS "\n"
	       "Compiler path: " CFUN_COMPILER "\n"
#ifdef CFUNGE_COMP_CLANG
	       "Compiler:      clang (unknown version)\n"
#elif defined(CFUNGE_COMP_ICC)
	       "Compiler:      ICC " FUNGE_CPP_STRINGIFY(__INTEL_COMPILER) "\n"
#elif defined(CFUNGE_COMP_GCC)
	       "Compiler:      GCC " FUNGE_CPP_STRINGIFY(__GNUC__)
	       "." FUNGE_CPP_STRINGIFY(__GNUC_MINOR__)
	       "." FUNGE_CPP_STRINGIFY(__GNUC_PATCHLEVEL__)
	       " (or compatible)\n"
#else
	       "Compiler:      Unknown.\n"
#endif
	       "Build type:    " CFUN_BUILD_TYPE "\n"
	       "Configured on: " CFUN_COMPILED_ON "\n\n"
	       "CFLAGS=\"" CFUN_USER_CFLAGS "\"\n"
	       "LDFLAGS=\"" CFUN_USER_LDFLAGS "\"\n",
	       sizeof(void*) * CHAR_BIT, sizeof(funge_cell) * CHAR_BIT);
	exit(EXIT_SUCCESS);
}

FUNGE_ATTR_NOINLINE FUNGE_ATTR_COLD FUNGE_ATTR_NORET
static void print_version(void)
{
	puts("cfunge " CFUNGE_APPVERSION "\n"
	     "Copyright (C) 2008-2013 Arvid Norlander.\n"
	     "This is free software.  You may redistribute copies of it under the terms of\n"
	     "the GNU General Public License <http://www.gnu.org/licenses/gpl.html>.\n"
	     "There is NO WARRANTY, to the extent permitted by law.\n\n"
	     "Written by Arvid Norlander.");

	exit(EXIT_SUCCESS);
}

int main(int argc, char *argv[])
{
	int opt;

#ifdef FUZZ_TESTING
	struct rlimit limit;
	alarm(3);
	getrlimit(RLIMIT_AS, &limit);
	limit.rlim_cur = 1024 * 1024 * 1024; // 1 GB
	setrlimit(RLIMIT_AS, &limit);
#endif

	// We detect socket issues in other ways.
	signal(SIGPIPE, SIG_IGN);

	while ((opt = getopt(argc, argv, "+bEFfhSs:t:VvW")) != -1) {
		switch (opt) {
			case 'b':
				setvbuf(stdout, cfun_iobuf, _IOFBF, sizeof(cfun_iobuf));
				break;
			case 'E':
				setting_enable_errors = true;
				break;
			case 'F':
				setting_disable_fingerprints = true;
				break;
			case 'f':
				print_features();
				break;
			case 'h':
				print_help();
				break;
			case 'S':
				setting_enable_sandbox = true;
				break;
			case 's':
				if (strncmp(optarg, "93", 2) == 0)
					setting_current_standard = stdver93;
				else if (strncmp(optarg, "98", 2) == 0)
					setting_current_standard = stdver98;
				else if (strncmp(optarg, "109", 3) == 0)
					setting_current_standard = stdver109;
				else {
					diag_fatal_format("%s is not valid for -s.\n", optarg);
				}
				break;
			case 't':
				setting_trace_level = (uint_fast16_t)atoi(optarg);
				break;
			case 'V':
				print_version();
				break;
			case 'v':
				print_build_info();
				break;
			case 'W':
				setting_enable_warnings = true;
				break;
			default:
				fprintf(stderr, "For help see: %s -h\n", argv[0]);
				return EXIT_FAILURE;
		}
	}
	if (FUNGE_UNLIKELY(optind >= argc)) {
		diag_fatal("No file provided.");
	} else {
		// Store argument count and a pointer to argv[optind] for later use
		// by the y instruction.
		fungeargc = argc - optind;
		fungeargv = (const char * const *)&argv[optind];
		// Run the actual interpreter (never returns).
		interpreter_run(argv[optind]);
	}
	// NEVER REACHED.
}
#endif /* ! CFUN_IS_IFFI */
